doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -525,31 +525,41 @@ This section describes some special purpose tactics to work with
 
 .. tacn:: discriminate {? @induction_arg }
 
-   Proves any goal for which a hypothesis in the form :n:`@term__1 = @term__2`
-   states an impossible structural equality for an inductive type.
-   If :n:`@induction_arg` is not given, it checks all the hypotheses
-   for impossible equalities.
-   For example, :g:`(S (S O)) = (S O)` is impossible.
-   If provided, :n:`@induction_arg` is a proof of an equality, typically
-   specified as the name of a hypothesis.
-
-   If no :n:`@induction_arg` is provided and the goal is in the form
-   :n:`@term__1 <> @term__2`, then the tactic behaves like
-   :n:`intro @ident; discriminate @ident`.
+   Proves goals for which a hypothesis or a :term:`premise` in
+   the goal that is convertible to the form :n:`@term__1 = @term__2`
+   has inconsistent constructors between the two sides of
+   the equality (i.e., a contradiction).  The tactic also proves goals
+   in the form :n:`@term__1 <> @term__2` that are inconsistent
+   (:ref:`example <discriminate_goal_inequality_ex>`).
+
+   The tactic relies on the fact that constructors of inductive types are injective
+   and disjoint, i.e. if `C1` and `C2` are distinct constructors of an inductive type then
+   :n:`C1 @term__1 = C1 @term__2` implies that :n:`@term__1 = @term__2` (injectivity)
+   and :n:`C1 @term__1 = C2 @term__2` is a contradiction (disjointedness).
+   For example, :g:`S (S O) = S O` is a contradiction: while
+   the outermost constructors are both `S`, the next ones differ (`S` versus `O`).
 
    The tactic traverses the normal forms of :n:`@term__1` and :n:`@term__2`,
```

Comment:
```suggestion
   The tactic traverses the head normal forms of :n:`@term__1` and :n:`@term__2`,
```

Reply:
But is "head normal form" accurate?  I thought the head normal form of `2 + 3` is `S (1 + 3)`, but I'm imagining discriminate still handles `2 + 3 = 6`?

Reply:
Ah, I guess maybe it only ever computes head normal forms, but it does that on sub terms, so we end up walking the normal form

Reply:
Yes, it walks the normal form up to the position where it differs. But after all, since we get a contradiction, we cannot really observe what processus is used to find the positions that differ, so, whatever explanation is probably fine.

As a side remark, it matters however in the case of `injection`. There is actually even an extra call to `simpl` on the resulting equations, subterms included, which looks too strong for what is worth.